src/inmanta/deploy/state.py

```diff
@@ -139,25 +197,119 @@ class ModelState:
     """
 
     version: int
-    resources: dict[ResourceIdStr, ResourceDetails] = dataclasses.field(default_factory=dict)
+    resources: dict["ResourceIdStr", ResourceDetails] = dataclasses.field(default_factory=dict)
     requires: RequiresProvidesMapping = dataclasses.field(default_factory=RequiresProvidesMapping)
-    resource_state: dict[ResourceIdStr, ResourceState] = dataclasses.field(default_factory=dict)
+    resource_state: dict["ResourceIdStr", ResourceState] = dataclasses.field(default_factory=dict)
     # resources with a known or assumed difference between intent and actual state
-    dirty: set[ResourceIdStr] = dataclasses.field(default_factory=set)
+    # (might be simply a change of its dependencies), which are still being processed by
+    # the resource scheduler. This is a short-lived transient state, used for internal concurrency control. Kept separate from
+    # ResourceStatus so that it lives outside the scheduler lock's scope.
+    dirty: set["ResourceIdStr"] = dataclasses.field(default_factory=set)
     # types per agent keeps track of which resource types live on which agent by doing a reference count
     # the dict is agent_name -> resource_type -> resource_count
-    types_per_agent: dict[str, dict[ResourceType, int]] = dataclasses.field(
+    types_per_agent: dict[str, dict["ResourceType", int]] = dataclasses.field(
         default_factory=lambda: defaultdict(lambda: defaultdict(lambda: 0))
     )
 
+    @classmethod
+    async def create_from_db(
+        cls, environment: uuid.UUID, last_processed_model_version: int, *, connection: asyncpg.connection.Connection
+    ) -> "ModelState":
+        """
+        Create a new instance of the ModelState object, by restoring the model state from the database.
+
+        :param environment: The environment the model state belongs to.
+        :param last_processed_model_version: The model version that was last processed by the scheduler, i.e. the version
+                                             associated with the state in the persistent_resource_state table.
+        """
+        from inmanta import data, resources
+
+        result = ModelState(version=last_processed_model_version)
+        resource_records = await data.Resource.get_resources_for_version_raw_with_persistent_state(
+            environment=environment,
+            version=last_processed_model_version,
+            projection=["resource_id", "attributes", "attribute_hash"],
+            projection_persistent=[
+                "is_orphan",
+                "is_undefined",
+                "current_intent_attribute_hash",
+                "last_deployed_attribute_hash",
+                "deployment_result",
+                "blocked_status",
+                "last_success",
+                "last_produced_events",
+            ],
+            project_attributes=["requires", const.RESOURCE_ATTRIBUTE_SEND_EVENTS],
+            connection=connection,
+        )
```

Comment:
given the amount of data we need to massage below: why not use the DTO?

Reply:
This method joins the resource table with the resource_persistent_state table and only selects very specific columns from the result. If we would go for the proposed approach of using the DTO we would:
1. Either have to make a lot more queries to the database.
2. Or, return all columns from the database.
Given that this method loads all the resources in the latest released version, I don't like option 1. Option 2 is possible if we accept the performance penalty. Wdyt?

Reply:
Probably fine!